<img src="images/banner_friends.jpg" width=80%>

<a name='outline'></a>

## Outline of the Project

This project has seven sections. Use the outline below to help you quickly navigate to the part of the project you're working on. Questions are worth one point each, unless they contain a ⭐️⭐️ next to them, in which case they are worth two points (e.g. **Question 1.2. ⭐️⭐️**). In most cases, questions worth two points are longer and more challenging than questions worth one point, though sometimes two point questions are themselves straightforward, but used to test the correctness of a previously implemented function.

-  [Welcome 👋](#welcome)
-  [About the Show 📺](#about_show)
-  [The Reboot🆕](#reboot)
-  [About the Data 💾](#about_data)
-  Section 1: [The Pilot 🎬](#section1)
-  Section 2: [The One with the Best Director 🏆](#section2)   
-  Section 3: [The One About Gender Balance 👩⚖️🧑](#section3)  
-  Section 4: [The Emotional One 😁😢😨](#section4)
-  Section 5: [The One with the Spoilers 🙊](#section5)
-  Section 6: [The One with the Generated Episode Titles 🖨️](#section6)
-  Section 7: [The Last One 🔚🥳](#section7)

<a id='welcome'></a>
## Welcome 👋

Welcome to the Final Project! Projects in DSC 10 are similar in format to homeworks, but are different in a few key ways. First, a project is comprehensive, meaning that it draws upon everything we've learned this quarter so far. Second, since problems can vary quite a bit in difficulty, some problems will be worth more points than others. Finally, in a project, the problems are more open-ended; they will usually ask for some result, but won't tell you what method should be used to get it. There might be several equally-valid approaches, and several steps might be necessary. This is closer to how data science is done in "real life".

It is important that you **start early** on the project! It is the final assignment that is due this quarter, but it is due just a few days before the Final Exam. You are especially encouraged to **find a partner** to work through the project with. You can work with anyone from any section of the course, but you must follow the [Project Partner Guidelines](https://dsc10.com/project-partners/) on the course website. In particular, you are both required to actively contribute to all parts of the project, and you are not allowed to split up the problems and each work on certain problems. If working in a pair, you should submit one notebook to Gradescope for the both of you.

**Important:** The `otter` tests don't usually tell you that your answer is correct. More often, they help catch basic mistakes. It's up to you to ensure that your answer is correct. If you're not sure, ask someone (not for the answer, but for some guidance about your approach). Directly sharing answers between groups is not okay, but discussing problems with the course staff or with other students is encouraged.

Please do not import any additional packages - you don't need them, and our autograder may not be able to run your code if you do.

As you work through this project, there are a few resources you may want to have open:
- [`babypandas` Notes](https://notes.dsc10.com/front.html)
- [Course Textbook](https://inferentialthinking.com/chapters/intro.html)
- [DSC 10 Reference Sheet](https://drive.google.com/file/d/1mQApk9Ovdi-QVqMgnNcq5dZcWucUKoG-/view)
- [`babypandas` Documentation](https://babypandas.readthedocs.io/en/latest/)
- Other links in the [Resources](https://dsc10.com/resources/) and [Debugging](https://dsc10.com/debugging/) tabs of the course website

Start early, good luck, and let's get started! 😎

In [ ]:
# Don't change this cell; just run it.
import babypandas as bpd
import numpy as np

import matplotlib.pyplot as plt
from matplotlib.offsetbox import OffsetImage, AnnotationBbox
plt.style.use('ggplot')

import otter
grader = otter.Notebook()

<a id='about_show'></a>
## About the Show 📺

*Friends* is a TV sitcom that aired from 1994 to 2004. The show revolves around six main characters, a group of friends in their 20s and 30s who live in New York City 🗽🌇.  Viewers watch these characters' lives, careers, and relationships develop over a ten year period.

<img src="images/friends.png" width="500" height="auto">

The show was critically well-received and popular among viewers. It set a number of records including:
- the number one television show in 2002
- the fifth-most-watched series finale in television history
- the most-watched episode of the 2000s decade
- Outstanding Comedy Series Primetime Emmy award 🏆
- one of *TV Guide*'s 50 Greatest Shows of All Time

<img src="images/friends_award.png" width="500" height="auto">

*Friends* used to be available to stream on Netflix, but with its incredible success, the show became too expensive to continue offering access. The last time *Friends* was on Netflix in the US, [it cost Netflix $100 million dollars 💸](https://slate.com/culture/2018/12/netflix-friends-streaming-100-million.html) just to offer the show on their streaming platform for one year!

If you'd like to see the show for yourself, all 10 seasons are available on [MAX](https://www.max.com/shows/friends/52dae4c7-2ab1-4bb9-ab1c-8100fd54e2f9?utm_id=sa%7C71700000110368783%7C58700008392050425%7Cp78367425622&gad_source=1&gclid=CjwKCAiAuYuvBhApEiwAzq_YiRkizSXADw1h1V4etPEJoJ6vam8ozaJ5pO1m9XqTh4JFxCZFO_7skxoC9jIQAvD_BwE&gclsrc=aw.ds). 🎥

<a id='reboot'></a>
## The Reboot 🆕

As an avid fan of *Friends*, you have decided to make your own reboot of the classic sitcom. A reboot of a TV show is a new show that takes place in the same setting as an older show, perhaps updated in some ways. Some reboots have a brand new cast with just a few original characters, meanwhile some reboots get the whole cast of the original show to star in the new version. For example, the reboot *How I Met Your Father* is based on the original *How I Met Your Mother*. Other examples include *That 90's Show*, a reboot of *That 70's Show*, and *Fuller House*, a reboot of *Full House*.

<img src="images/full_house.png" width="500" height="auto">

You've noticed that sitcom reboots are on the rise and you want to get in on the action by finding all the best and most successful aspects of *Friends* and turning them into an even better version. In order to make the reboot, you'll have to do a lot of planning and research to present your plan to the production company that will fund the show. Here's a list of all the things you'll want to have prepared by the time you meet with the producers. Each item on the list represents a section of this project.

1. Create a list of writers and directors to contact.
2. Select a top choice for the director of the reboot.
3. Determine whether to assign spoken lines to male and female characters equally.
4. Make a visualization showing the emotional range for each main character to present to the writers.
5. Decide whether the main characters should be romantically involved with one another.
6. Make a list of episode titles for the first season of the new show.
7. Decide how many episodes the reboot should have, based on viewership and rating data.

After you've done all of these things you will have completed your plan for the reboot!

<a id='about_data'></a>
## About the Data 💾

Our data, [originally from Kaggle](https://www.kaggle.com/datasets/sujaykapadnis/friends?select=friends.csv), includes several different CSVs about *Friends* episodes, lines, and emotions. If you find any errors in the data, do not attempt to fix them; just analyze the data you are given.

### Episodes 🎬

The `episodes` DataFrame has a row for each episode from all 10 seasons of *Friends*. For each episode, we have the
- season number (`'season'`)
- episode number (`'episode'`)
- title (`'title'`)
- director(s) (`'directed_by'`)
- writer(s) (`'written_by'`)
- premiere date (`'air_date'`)
- number of views on premiere date in millions (`'us_views_millions'`)
- rating out of ten (`'imdb_rating'`)

For example, the first episode of the first season is called `'The Pilot'`. It was directed by James Burrows and written by David Crane and Marta Kauffman. When it aired on September 22nd, 1994, there were 21.5 million viewers, and it was rated 8.3/10 on IMDb.

Run the cell below to load in the data.

In [ ]:
episodes = bpd.read_csv('data/friends_info.csv')
episodes

season  episode                                           title  \
0         1        1                                       The Pilot   
1         1        2            The One with the Sonogram at the End   
2         1        3                          The One with the Thumb   
3         1        4              The One with George Stephanopoulos   
4         1        5  The One with the East German Laundry Detergent   
..      ...      ...                                             ...   
231      10       14                  The One with Princess Consuela   
232      10       15                      The One Where Estelle Dies   
233      10       16          The One with Rachel's Going Away Party   
234      10       17                                    The Last One   
235      10       18                                    The Last One   

         directed_by                                         written_by  \
0      James Burrows                       David Crane & Marta Kauffman   
1      James Burrows                       David Crane & Marta Kauffman   
2      James Burrows                     Jeffrey Astrof & Mike Sikowitz   
3      James Burrows                                        Alexa Junge   
4      Pamela Fryman                     Jeff Greenstein & Jeff Strauss   
..               ...                                                ...   
231   Gary Halvorson  Story by: Robert CarlockTeleplay by: Tracy Reilly   
232   Gary Halvorson  Story by: Mark KunerthTeleplay by: David Crane...   
233   Gary Halvorson                           Andrew Reich & Ted Cohen   
234  Kevin S. Bright                       Marta Kauffman & David Crane   
235  Kevin S. Bright                       Marta Kauffman & David Crane   

     air_date  us_views_millions  imdb_rating  
0     9/22/94              21.50          8.3  
1     9/29/94              20.20          8.1  
2     10/6/94              19.50          8.2  
3    10/13/94              19.70          8.1  
4    10/20/94              18.60          8.5  
..        ...                ...          ...  
231   2/26/04              22.83          8.6  
232   4/22/04              22.64          8.5  
233   4/29/04              24.51          8.9  
234    5/6/04              52.46          9.7  
235    5/6/04              52.46          9.7  

[236 rows x 8 columns]

### Lines 🗣

While the `episodes` DataFrame will give you plenty of information on the writers and directors of the show and the success of each episode, you want to know more about the characters and their interactions on the show. You have detailed line-by-line scripts for the dialogue and scene directions of 30 randomly selected *Friends* episodes, stored in a DataFrame called `lines`. For each spoken line or scene direction of these 30 episodes, the `lines` DataFrame includes the following columns:
- season number (`'season'`)
- episode number (`'episode'`)
- scene number (`'scene'`)
- line number (`'utterance'`)
- character that said the line (`'speaker'`), which is sometimes just `'Scene Directions'`
- text of the line (`'text'`)

For example, the seventh episode of season one opens with character Rachel Green saying `
'Everybody? Shh, shhh. Uhhh... Central Perk is proud to present the music of Miss Phoebe Buffay.'`

Run the cell below to load in the data.

In [ ]:
lines = bpd.read_csv('data/friends_sample.csv')
lines

text           speaker  \
0     Everybody? Shh, shhh. Uhhh... Central Perk is ...      Rachel Green   
1                                            (applause)  Scene Directions   
2     Hi. Um, I want to start with a song thats abou...     Phoebe Buffay   
3                            Oh, great. This is just...     Chandler Bing   
4     (Chandler sees that there is a gorgeous model ...  Scene Directions   
...                                                 ...               ...   
8452                                     That I can do.    Joey Tribbiani   
8453  Come on! You can drink a gallon of milk in 10 ...     Phoebe Buffay   
8454     All right, watch me! Okay, you time me. Ready?    Joey Tribbiani   
8455                                       Ready... GO!     Phoebe Buffay   
8456                                        You did it!     Phoebe Buffay   

      season  episode  scene  utterance  
0          1        7      1          1  
1          1        7      1          2  
2          1        7      1          3  
3          1        7      2          1  
4          1        7      2          2  
...      ...      ...    ...        ...  
8452      10       13     13          9  
8453      10       13     13         10  
8454      10       13     13         11  
8455      10       13     13         12  
8456      10       13     13         13  

[8457 rows x 6 columns]

### Emotions

Lastly, you have a dataset of the emotion attached to each line of the show, for most (but not all) lines in the first four seasons of the show. For each line included in our dataset, the line is associated with one of seven different emotions. These seven emotions are:
- `'Joyful'` 😄
- `'Mad'` 😡
- `'Sad'` 😭
- `'Powerful'` 💪
- `'Peaceful'` ✌
- `'Scared'` 😱
- `'Neutral'` 😑

Classifying each line's emotion is a huge data science task in and of itself. [This paper](https://arxiv.org/pdf/1708.04299) explains how convolutional neural networks were able to do this task, if you're interested in learning more.

Run the cell below to load in the data, which has the same columns as `lines` plus an extra column called `emotion`. For example, the first line of the fourth scene in Season 1, Episode 1 is categorized as `'Mad'`. That means the speaker, Ross, was probably mad when he said it. If you've ever assembled furniture, you might understand this frustration.

In [ ]:
emotions = bpd.read_csv('data/friends_emotions.csv')
emotions

text         speaker  \
0      I'm supposed to attach a brackety thing to the...     Ross Geller   
1                I'm thinking we've got a bookcase here.  Joey Tribbiani   
2                                It's a beautiful thing.   Chandler Bing   
3                                           What's this?  Joey Tribbiani   
4      I would have to say that is an 'L'-shaped brac...   Chandler Bing   
...                                                  ...             ...   
12601        Ahh, yes, I will have a glass of the Merlot    Rachel Green   
12602                                              Okay.     Air Hostess   
12603        And uh, he will have a white wine spritzer.    Rachel Green   
12604  Okay, good. Thank you. I'll be back shortly, a...     Air Hostess   
12605  All right. Woo! Hey, look at that, the airport...    Rachel Green   

       season  episode  scene  utterance  emotion  
0           1        1      4          1      Mad  
1           1        1      4          3  Neutral  
2           1        1      4          4   Joyful  
3           1        1      4          5  Neutral  
4           1        1      4          6  Neutral  
...       ...      ...    ...        ...      ...  
12601       4       24     25          2  Neutral  
12602       4       24     25          3  Neutral  
12603       4       24     25          4  Neutral  
12604       4       24     25          5   Joyful  
12605       4       24     25          6   Scared  

[12606 rows x 7 columns]

<a id='section1'></a>
## Section 1: The Pilot 🎬

<small>(<a href="#outline">return to the outline</a>)</small>

To start, we'll try to find the most experienced writers and directors in the show to know who to contact for making the reboot. The `episodes` DataFrame contains the writers and directors for each episode, but it's not in the neatest format.

In [ ]:
episodes

Some episodes were written or directed by more than one person, and many episodes have certain people who wrote the story and other people who wrote the teleplay. So identifying the most experienced writers and directors is not as straightforward as simply grouping and counting. First, we'll have to clean up the data to break up each string of names into a list of individual names.

**Question 1.1.**
We want to first handle any episodes with two directors so that later we can count the number of episodes each director contributed to. Create a function called `parse_names` that turns a string in the `'directed_by'` column to a **list** of one or two names. Values in the `'directed_by'` column are either single names, or two names separated by an ampersand. For example,
-  `parse_names('James Burrows')` should return `['James Burrows']`, and
-  `parse_names('Kevin S. Bright & Gary Halvorson')` should return `['Kevin S. Bright', 'Gary Halvorson']`.

Then, apply your function to the `'directed_by'` column from the `episodes` DataFrame, and store the resulting Series in the variable `directors_by_episode`. **Do not** modify the `episodes` DataFrame.

**Question 1.1.**
We want to first handle any episodes with two directors so that later we can count the number of episodes each director contributed to. Create a function called `parse_names` that turns a string in the `'directed_by'` column to a **list** of one or two names. Values in the `'directed_by'` column are either single names, or two names separated by an ampersand. For example,
-  `parse_names('James Burrows')` should return `['James Burrows']`, and
-  `parse_names('Kevin S. Bright & Gary Halvorson')` should return `['Kevin S. Bright', 'Gary Halvorson']`.

Then, apply your function to the `'directed_by'` column from the `episodes` DataFrame, and store the resulting Series in the variable `directors_by_episode`. **Do not** modify the `episodes` DataFrame.

<h1>.apply is equivalent to each value being choosen loc wise<h1>


In [ ]:
names

array(['James Burrows', 'James Burrows', 'James Burrows', 'James Burrows',
       'Pamela Fryman', 'Arlene Sanford', 'James Burrows',
       'James Burrows', 'James Burrows', 'Peter Bonerz', 'James Burrows',
       'Paul Lazarus', 'Alan Myerson', 'James Burrows', 'Alan Myerson',
       'Michael Lembeck', 'Michael Lembeck', 'James Burrows',
       'Peter Bonerz', 'Peter Bonerz', 'Gail Mancuso', 'Robby Benson',
       'James Burrows', 'Kevin S. Bright', 'Michael Lembeck',
       'Michael Lembeck', 'Kevin S. Bright', 'Gail Mancuso',
       'Ellen Gittelsohn', 'Gail Mancuso', 'Peter Bonerz',
       'Mary Kay Place', 'Kevin S. Bright', 'Thomas Schlamme',
       'Thomas Schlamme', 'Michael Lembeck', 'Michael Lembeck',
       'James Burrows', 'Michael Lembeck', 'Michael Lembeck',
       'Michael Lembeck', 'Michael Lembeck', 'Michael Lembeck',
       'Michael Lembeck', 'Michael Lembeck', 'Michael Lembeck',
       'Michael Lembeck', 'Michael Lembeck', 'Gail Mancuso',
       'Gail Mancuso', 'Kev

In [ ]:
names = (episodes.get("directed_by").values)

for entry in names:
    for i in entry:
        if i == "&":
            print(entry)
        else:
            continue
    print(entry)

James Burrows
James Burrows
James Burrows
James Burrows
Pamela Fryman
Arlene Sanford
James Burrows
James Burrows
James Burrows
Peter Bonerz
James Burrows
Paul Lazarus
Alan Myerson
James Burrows
Alan Myerson
Michael Lembeck
Michael Lembeck
James Burrows
Peter Bonerz
Peter Bonerz
Gail Mancuso
Robby Benson
James Burrows
Kevin S. Bright
Michael Lembeck
Michael Lembeck
Kevin S. Bright
Gail Mancuso
Ellen Gittelsohn
Gail Mancuso
Peter Bonerz
Mary Kay Place
Kevin S. Bright
Thomas Schlamme
Thomas Schlamme
Michael Lembeck
Michael Lembeck
James Burrows
Michael Lembeck
Michael Lembeck
Michael Lembeck
Michael Lembeck
Michael Lembeck
Michael Lembeck
Michael Lembeck
Michael Lembeck
Michael Lembeck
Michael Lembeck
Gail Mancuso
Gail Mancuso
Kevin S. Bright
Steve Zuckerman
Steve Zuckerman
Peter Bonerz
Gail Mancuso
Gail Mancuso
Kevin S. Bright
Terry Hughes
Terry Hughes
Robby Benson
Robby Benson
Robby Benson
James Burrows
James Burrows
Sam Simon
Robby Benson
Terry Hughes
Terry Hughes
Michael Lembeck
Peter

<h1>this loop created duplicate entries at "&" results and other loc(s) why :(<h1>

In [ ]:
def parse_names(names):
    '''Returns a list of individual names present in the input string.'''
    separated_names = np.array([])


    if names.contains()
    names_per_episode = (names).split('&')
    for i in names_per_episode:
        if (i) == "Award-Winning":
            return "Award-Winning"
    return ("No Award")

directors_by_episode = episodes.get("directed_by").apply(parse_names)
directors_by_episode



def label_award(genres):
    genre_per_show = (genres).split(', ')
    for i in genre_per_show:
        if (i) == "Award-Winning":
            return "Award-Winning"
    return ("No Award")


In [ ]:
def parse_names(names):
    '''Returns a list of individual names present in the input string.'''
    separated_names = []
    # thus the range of this loop must be the length of the entry \
    # word in the array NOT the length of the whole column
    for entry in names.values:
        for i in np.arange(len(names[i])):
                separated_names = np.append(separated_names, list(names))

                separated_names = np.append(separated_names, list(names)[i].split(" "))
    return
# (episodes.get("directed_by")).apply(parse_names)
parse_names(episodes.get("directed_by"))
# directors_by_episode =
# directors_by_episode

In [ ]:
grader.check("q1_1")


For the next question, you'll need to know something interesting about how lists work in Python: when you sum two lists together, the output is one giant list that contains all the elements in both lists combined. An example is shown below.


In [ ]:
['List', 'combining'] + ['is', 'my', 'passion']

**Question 1.2.** ⭐⭐
Now it's time to count the number of episodes each director directed. Make a DataFrame called `directors`, indexed by `'name'`, with one column called `'num_episodes'`. There should be one row for each unique director, and `'num_episodes'` should contain the number of episodes they directed or co-directed. Sort this DataFrame so that the most experienced directors appear at the top.

***Hints:***
- Our solution involved creating a new, empty DataFrame with `bpd.DataFrame()`, adding columns, and using `groupby`.
- For `groupby` to give meaningful results, you must use it on a DataFrame with at least two columns.

In [ ]:
directors = ...
directors

In [ ]:
grader.check("q1_2")

Now we want to repeat the process we've done so far in this section, except for writers instead of directors. This is a little trickier because for some episodes, certain writers wrote the story, meaning they developed the plot of the episode, while others wrote the teleplay, which means they wrote the dialogue and scene directions based on the story. For other episodes, the same writers wrote both the story and teleplay.

There are two ways writers are represented in the `'written_by'` column:

1. One or two names separated by `&` (similar to the `'directed_by'` column). This means these writers wrote both the story and teleplay.
2. One or two names separated by `&` after `Story by: ` and one or two names separated by `&` after `Teleplay by: `

Therefore, the six possible templates of strings in the `'written_by'` column are:
1. <ins>writer</ins>
2. <ins>writer</ins> & <ins>other writer</ins>
3. Story by: <ins>writer</ins>Teleplay by: <ins>other writer</ins>
4. Story by: <ins>writer</ins>Teleplay by: <ins>other writer</ins> & <ins>another writer</ins>
5. Story by: <ins>writer</ins> & <ins>other writer</ins>Teleplay by: <ins>another writer</ins>
6. Story by: <ins>writer</ins> & <ins>other writer</ins>Teleplay by: <ins>another writer</ins> & <ins>yet another writer</ins>

**Question 1.3.** Implement the function `get_teleplay_writers`, which takes as input a string from the `'written_by'` column, corresponding to one episode, and returns a list of teleplay writers for that episode.  Remember that if the writers' names aren't divided into story and teleplay, then all writers wrote both the story and the teleplay.

***Hints:***
- Use the `parse_names` function you wrote earlier.
- While there are six templates, you can solve this problem without using any conditional logic (`if`-statements). Your solution doesn't have to be the same as ours, but ours was just one line!

In [ ]:
def get_teleplay_writers(names):
    '''Returns a list of names of teleplay writers present in the input string.'''
    ...

In [ ]:
grader.check("q1_3")

**Question 1.4.** Now we need a function for getting the names of the writers that wrote the story of an episode. Create a function `get_story_writers` that takes a string from the `'written_by'` column and returns a list of story writers. As before, if the writers' names aren't divided into story and teleplay, then all writers wrote both the story and the teleplay.


In [ ]:
def get_story_writers(names):
    '''Returns a list of names of story writers present in the input string.'''
    ...

In [ ]:
grader.check("q1_4")

**Question 1.5.**
Next we want to count how many times each writer wrote or co-wrote the story for an episode, as well as how many times each writer wrote or co-wrote the teleplay for an episode. This will be similar to the process we did in Question 1.2 for directors. Since we want to do something similar for story writers and teleplay writers, let's create a more general function that works for directors, story writers, and teleplay writers.

The function `count_episodes` should take as input a Series of lists, containing for each episode, either the directors, story writers, or teleplay writers for that episode. The function should return a DataFrame indexed by `'name'`, with one column called `'num_episodes'` containing a count of how many episodes each staff member worked on, sorted in descending order of the number of episodes.

For example `count_episodes(directors_by_episode)` should return a DataFrame that is identical to the `directors` DataFrame you created in Question 1.2. But we could also use the same function, with different inputs, to count the number of episodes each story writer and each teleplay writer worked on.

In [ ]:
def count_episodes(staff):
    '''Returns a DataFrame showing how many episodes each staff member worked on.'''
    ...

# An example call to your function. Feel free to change this and try out other inputs.
count_episodes(directors_by_episode)

In [ ]:
grader.check("q1_5")

**Question 1.6.**
Now, use the function you wrote in Question 1.5 to make a DataFrame called `story_writers`. This DataFrame should be indexed by `'name'`, with one column called `'num_episodes'`. There should be one row for each unique story writer, and `'num_episodes'` should contain the number of episodes for which they wrote the story, either independently or with a co-writer. Sort this DataFrame so that the most experienced story writers appear at the top.

Then do the same for teleplay writers, storing your result in `teleplay_writers`.

In [ ]:
story_writers = ...
story_writers

In [ ]:
grader.check("q1_6_a")

In [ ]:
teleplay_writers= ...
teleplay_writers

In [ ]:
grader.check("q1_6_b")

**Question 1.7.** ⭐⭐
Finally, it's time to use the information in `directors`, `story_writers`, and `teleplay_writers` to find the best candidates to help create the reboot. Create an array called `experienced_directors` containing the names of directors that directed at least 20 episodes. Then create an array called `experienced_writers` containing the names of writers that wrote at least 20 stories **and** at least 20 teleplays.

In [ ]:
experienced_directors = ...
experienced_writers = ...

print(f'The experienced directors are {experienced_directors}')
print(f'The experienced writers are {experienced_writers}')

In [ ]:
grader.check("q1_7")

These are the writers and directors you'll reach out to about your reboot, and hopefully, you can find at least one writer and one director interested in working with you on the reboot! If not, you can always relax your criteria and ask some less-experienced writers and directors.

<a id='section2'></a>
## Section 2: The One with the Best Director 🏆

<small>(<a href="#outline">return to the outline</a>)</small>

You discovered in Section 1 that the two directors who directed the most episodes were Gary Halvorson and Kevin S. Bright. In this section, we want to discover which of these two directors makes better episodes, as determined by viewers. We'll use the `'imdb_rating'` column from the `episodes` DataFrame to gauge how much people enjoyed an episode.

**Question 2.1.** Create a DataFrame `mean_by_director` with the director's name as the index and just one column called `'mean_rating'` that contains the mean rating of all of the episodes directed by that director.

***Note:*** There is one episode that had two directors, separated by `'&'`. Exclude this episode, so that we're only looking at episodes with a single director.

In [ ]:
episodes

season  episode                                           title  \
0         1        1                                       The Pilot   
1         1        2            The One with the Sonogram at the End   
2         1        3                          The One with the Thumb   
3         1        4              The One with George Stephanopoulos   
4         1        5  The One with the East German Laundry Detergent   
..      ...      ...                                             ...   
231      10       14                  The One with Princess Consuela   
232      10       15                      The One Where Estelle Dies   
233      10       16          The One with Rachel's Going Away Party   
234      10       17                                    The Last One   
235      10       18                                    The Last One   

         directed_by                                         written_by  \
0      James Burrows                       David Crane & Marta Kauffman   
1      James Burrows                       David Crane & Marta Kauffman   
2      James Burrows                     Jeffrey Astrof & Mike Sikowitz   
3      James Burrows                                        Alexa Junge   
4      Pamela Fryman                     Jeff Greenstein & Jeff Strauss   
..               ...                                                ...   
231   Gary Halvorson  Story by: Robert CarlockTeleplay by: Tracy Reilly   
232   Gary Halvorson  Story by: Mark KunerthTeleplay by: David Crane...   
233   Gary Halvorson                           Andrew Reich & Ted Cohen   
234  Kevin S. Bright                       Marta Kauffman & David Crane   
235  Kevin S. Bright                       Marta Kauffman & David Crane   

     air_date  us_views_millions  imdb_rating  
0     9/22/94              21.50          8.3  
1     9/29/94              20.20          8.1  
2     10/6/94              19.50          8.2  
3    10/13/94              19.70          8.1  
4    10/20/94              18.60          8.5  
..        ...                ...          ...  
231   2/26/04              22.83          8.6  
232   4/22/04              22.64          8.5  
233   4/29/04              24.51          8.9  
234    5/6/04              52.46          9.7  
235    5/6/04              52.46          9.7  

[236 rows x 8 columns]

In [ ]:
episodes.get("directed_by").str.contains("e")

0       True
1       True
2       True
3       True
4       True
       ...  
231    False
232    False
233    False
234     True
235     True
Name: directed_by, Length: 236, dtype: bool

In [ ]:
mean_by_director = ...
mean_by_director

In [ ]:
grader.check("q2_1")

Now let's look at how Gary Halvorson's episodes are rated, on average.

In [ ]:
mean_by_director.get('mean_rating').loc['Gary Halvorson']

How accurate is this number in reflecting Gary's true potential as a director? In this section, we'll consider the data we have to be a sample of each director's potential abilities. Based on this sample, we want to infer each director's ability more broadly.

For example, the average rating of episodes that Gary actually directed was 8.4, but the set of episodes he *might* have directed with additional opportunities (representing his potential as a director) could have had a higher or lower average rating. By looking at a data set of *Friends* episodes directed by Gary, we are only looking at a sample from a larger population.

We'll estimate each director's mean episode rating in the population based on their mean episode in the sample using **bootstrapping**.

**Question 2.2.** Below, write a function called `simulate_estimates`. It should take 3 arguments:
- `director_df`: A DataFrame with a row for each episode, which includes columns `'directed_by'` and `'imdb_rating'`.
- `director`: The name of the director whose mean episode rating we are trying to estimate.
- `repetitions`: The number of repetitions to perform (the number of resamples to create).

It should take `repetitions` resamples with replacement from the rows of `director_df` that correspond to the given `director`. For each of those resamples, it should compute the mean episode rating for that resample. Then it should return an array containing the values of those means for each resample.

In [ ]:
def simulate_estimates(director_df, director, repetitions):
    '''Returns an array of length repetitions,
    containing bootstrapped means of the data
    in the imbdb_rating column for the given director. '''
    ...

In [ ]:
grader.check("q2_2")

**Question 2.3.** Use your function `simulate_estimates` to estimate the mean rating of Gary Halvorson's episodes. Use `repetitions = 5000`, and save your array of bootstrapped means in the variable `gary_boot_means`. Repeat the same process for Kevin S. Bright, storing your results in `kevin_boot_means`.  

Then, plot the distributions of these arrays in one [overlaid density histogram](https://dsc10.com/resources/lectures/lec07/lec07.html#Plotting-overlaid-histograms). Use the optional parameter `alpha=0.5` in your call to `.plot`, which changes the opacity so you can see the distributions more clearly.

In [ ]:
gary_boot_means = ...
kevin_boot_means = ...

# Plot your overlaid histogram here.
...

In [ ]:
grader.check("q2_3")

**Question 2.4.** Now we want to calculate a 99% confidence interval for the mean episode rating of each of the two directors. To do this, create a function `confidence_interval_99`, which takes in an array of bootstrapped statistics `boot_stats` and returns a list of length two, containing the left endpoint and the right endpoint of the 99% confidence interval.

In [ ]:
def confidence_interval_99(boot_stats):
    '''Returns a list of the endpoints of a 99% confidence interval based on boot_stats.'''
    ...

print('Gary 99% CI:', confidence_interval_99(gary_boot_means))
print('Kevin 99% CI:', confidence_interval_99(kevin_boot_means))

In [ ]:
grader.check("q2_4")

From what we've done so far, we've established that Kevin's episodes are generally rated better than Gary's episodes, however there is some overlap in their confidence intervals. This means it could be the case that Gary is just as strong a director as Kevin, but this is just not reflected in our sample.

We'll address this possibility by performing a hypothesis test with the following hypotheses:

- **Null Hypothesis:** The mean rating of Kevin's episodes in the population equals the mean rating of Gary's episodes in the population. Equivalently, the difference in the mean rating for Kevin's and Gary's episodes in the population equals 0.
- **Alternative Hypothesis:** The mean rating of Kevin's episodes in the population does not equal the mean rating of Gary's episodes in the population. Equivalently, the difference in the mean rating for Kevin's and Gary's episodes in the population does not equal 0.

Remember, the population represents all episodes they *might* have directed, or their potential as a director.

Since we were able to set up our hypothesis test as a question of whether our population parameter – the difference in mean rating for Kevin's and Gary's episodes – is equal to a certain value, we can **test our hypotheses by constructing a confidence interval for the parameter**. This is the method we used in [Lecture 22](https://dsc10.com/resources/lectures/lec22/lec22.html#Visualizing-the-distribution-of-each-group) to test whether the mean human body temperature was actually 98.6 degrees Fahrenheit. For a refresher on this method, you can read more about conducting a hypothesis test with a confidence interval in [CIT 13.4](https://inferentialthinking.com/chapters/13/4/Using_Confidence_Intervals.html#).

**Question 2.5.** ⭐⭐ Compute 1000 bootstrapped estimates for the difference in the mean rating for Kevin's episodes and Gary's episodes in the population (subtract in the order Kevin minus Gary). Store your 1000 estimates in the `difference_means` array.

You should generate your resamples of Kevin's episodes by sampling from the set of episodes directed by Kevin, and similarly for Gary, by sampling from the set of episodes directed by Gary.

In [ ]:
difference_means = ...

# Just display the first ten differences.
difference_means[:10]

In [ ]:
grader.check("q2_5")

Let's visualize your estimates:

In [ ]:
(bpd.DataFrame().assign(DifferenceMeans = difference_means)
 .plot(kind='hist', density=True, ec='w', figsize=(10, 5)));

**Question 2.6.** Use the function `confidence_interval_99` you created before to compute a 99% confidence interval for the difference in the mean rating of Kevin's and Gary's episodes (as before, Kevin minus Gary). Assign to `kevin_gary_difference_CI` a list containing the endpoints of this confidence interval.

In [ ]:
kevin_gary_difference_CI = ...
kevin_gary_difference_CI

In [ ]:
grader.check("q2_6")

Recall the hypotheses we were testing:

- **Null Hypothesis:** The mean rating of Kevin's episodes in the population equals the mean rating of Gary's episodes in the population. Equivalently, the difference in the mean rating for Kevin's and Gary's episodes in the population equals 0.
- **Alternative Hypothesis:** The mean rating of Kevin's episodes in the population does not equal the mean rating of Gary's episodes in the population. Equivalently, the difference in the mean rating for Kevin's and Gary's episodes in the population does not equal 0.

**Question 2.7.** Based on the confidence interval you've created, would you reject the null hypothesis at the 0.01 significance level? Set `reject_kevin_gary` to True if you would reject the null hypothesis, and False if you would not.

In [ ]:
reject_kevin_gary = ...
reject_kevin_gary

In [ ]:
grader.check("q2_7")

We have now discovered which of these two directors would likely make better episodes for your reboot. However, we also want to know whether Gary and Kevin's episodes have other differences besides the ratings they generate. For example, does one of them tend to direct episodes with more views?

Let's write a single function that works for both rating and viewership. To do this, we want to generalize our simulation code from Question 2.2 so that we can create a confidence interval for either variable.

**Question 2.8.** Create a function called `compare_gary_kevin`, which takes in 3 inputs:
- `sample_df`, a DataFrame with a row for each episode in our sample, which includes a column called `'directed_by'`.
- `variable`, the column name of the relevant variable whose mean we want to estimate.
- `repetitions`, the number of repetitions to perform (the number of resamples to create).
                       
The function should adhere to these specifications:
1. The function should generate an overlaid density histogram, showing the distributions of bootstrapped means of the given variable, for both Kevin and Gary. Make sure to give your histogram a descriptive title and to use appropriate labels. Use `bins=20`, `alpha=0.5`, and `figsize=(10,5)`.
2. The function should print a statement with the 99% confidence interval for the mean value of the given variable in the population, for both Kevin and Gary. See the example below for the type of statement to print, but the exact formatting is up to you.
3. The function should return nothing.

***Hints:***
- This is designed to be a challenging question, but remember that you can use any of the functions you've already created.
- Our solution does the necessary operations once for Gary and once for Kevin and assigns columns named `'Gary_mean_estimate'` and `'Kevin_mean_estimate'`.

Here is an example output that shows a comparison of estimates for Gary and Kevin's mean `'imdb_rating'`.
<img src="images/gary_kevin_example.jpg" width = 700>

In [ ]:
def compare_gary_kevin(sample_df, variable, repetitions):
    '''For each of Gary and Kevin, display a distribution of bootstrapped means and
    a confidence interval for the mean value of the variable from sample_df.'''
    ...

# Try to replicate the graph shown in the example.
compare_gary_kevin(episodes, 'imdb_rating', 1000)

**Question 2.9.** ⭐⭐ Using the `compare_gary_kevin` function you just wrote, create a plot and confidence intervals that would help you answer the following question.

>Whose episodes have more views on average, Kevin's or Gary's?

This question is not difficult; it is worth two points because it tests your implementation of the function `compare_gary_kevin` which you wrote in the previous question. All you need to do here is make one call to your function with the appropriate inputs. Use `repetitions=1000`.

<!-- BEGIN QUESTION -->

<!--
BEGIN QUESTION
name: q2_9
manual: True

-->

In [ ]:
# Make your function call here. Be sure to run this cell before submitting.
...

<!-- END QUESTION -->



At this point it should be clear which director you'll be reaching out to for your reboot!

<a id='section3'></a>
## Section 3: The One About Gender Balance 👩⚖️🧑

<small>(<a href="#outline">return to the outline</a>)</small>

After watching a couple of episodes of <i>Friends</i>, you start to wonder if the three male main characters (`'Ross Geller'`,
`'Chandler Bing'`, `'Joey Tribbiani'`) speak more lines than the the three female main characters (`'Rachel Green'`, `'Monica Geller'`, `'Phoebe Buffay'`). You want your reboot to be true to the spirit of the original show, but you don't feel great about creating a television show that has a significant gender imbalance.

Below is the `lines` DataFrame, which contains information on all the lines from a random sample of 30 *Friends* episodes. Notice that `lines` contains **every** line from each one of these 30 episodes, including lines from characters other than the main six, and even the scene directions.

In [ ]:
lines

**Question 3.1.** Write a function named `main_char` that takes in the name of a single speaker and determines whether that speaker is one of the six main characters:
- `'Ross Geller'`
- `'Chandler Bing'`
- `'Joey Tribbiani'`
- `'Rachel Green'`
- `'Monica Geller'`
- `'Phoebe Buffay'`

For example, `main_char('Chandler Bing')` should return `True`, but `main_char('Janice Hosenstein')` should return `False`.

Then, use your function to filter the `lines` DataFrame to contain only lines spoken by one of the six main characters, saving the result as `main_lines`.

In [ ]:
def main_char(speaker):
    '''Returns True if speaker is a main character, False otherwise.'''
    ...

main_lines = ...
main_lines

In [ ]:
grader.check("q3_1")

**Question 3.2.** Using the `main_lines` DataFrame, count the total number of lines spoken by a main character across all episodes in the sample and assign your answer to the variable `line_count`.

Then compute the proportion of these lines that were spoken by male characters (`'Ross Geller'`,
`'Chandler Bing'`, `'Joey Tribbiani'`) and the proportion of lines that were spoken by female characters (`'Rachel Green'`, `'Monica Geller'`, `'Phoebe Buffay'`). Assign your answers to the variables `observed_male_prop` and  `observed_female_prop`.

In [ ]:
line_count = ...
observed_female_prop = ...
observed_male_prop = ...

print('Number of Lines: ' + str(line_count))
print('Male Proportion: ' + str(observed_male_prop))
print('Female Proportion: ' + str(observed_female_prop))

In [ ]:
grader.check("q3_2")

You recognize that `observed_female_prop` and `observed_male_prop` are similar but they're not exactly the same. Is this just random chance that your sample happened to include more lines by male main characters? Or is the case that throughout the show *Friends*, lines spoken by one of the six main characters are actually more likely to be spoken by male characters? Let's do a hypothesis test with the following hypotheses:

- **Null Hypothesis**: Throughout the ten seasons of *Friends*, male main characters speak the same number of lines as female main characters.  
- **Alternative Hypothesis**: Throughout the ten seasons of *Friends*, male main characters speak more lines than female main characters.

Run the cell below to define a variable `null_distribution` that shows the proportion of each gender according to our model.

In [ ]:
null_distribution = np.array([0.5, 0.5])
null_distribution

**Question 3.3.** To perform our hypothesis test, we will simulate drawing a random sample of size `line_count` from the null distribution, and then compute a test statistic on each simulated sample. We must first choose a reasonable test statistic that will help us determine whether or not to reject the null hypothesis.

From the options below, find **all** valid test statistics that we could use for this hypothesis test. Save the numbers of your choices in a **list** called `gender_test_statistics`. Valid test statistics are ones that would allow us to distinguish between the null and alternative hypotheses.

***Hint:*** To determine whether a test statistic is valid, think about which values of the statistic (high, low, moderate) would make you lean towards the null and which would make you lean towards the alternative.

1. The difference between the proportion of female-spoken lines and 0.5.
2. The difference between the number of male-spoken lines and the number of female-spoken lines.
3. The difference between the number of female-spoken lines and one half of `line_count`.
4. Three times the difference between the proportion of male-spoken lines and 0.5.
5. The total variation distance between the gender distribution of lines spoken and the null distribution.

In [ ]:
gender_test_statistics = ...
gender_test_statistics

In [ ]:
grader.check("q3_3")

**Question 3.4.** For this hypothesis test, we'll use the proportion of male-spoken lines as our test statistic. Write a simulation that runs 10,000 times, each time drawing a random sample of size `line_count` under the assumption of the null hypothesis. Keep track of the simulated test statistics in the `gender_stats` array.

In [ ]:
gender_stats = ...

# Visualize with a histogram
bpd.DataFrame().assign(gender_stats=gender_stats).plot(kind='hist', density=True, ec='w', figsize=(10, 5));
plt.axvline(x=observed_male_prop, color='black', linewidth=4, label='observed statistic')
plt.legend();

In [ ]:
grader.check("q3_4")

**Question 3.5.** Compute the p-value for this hypothesis test, and save the result to `gender_p_value`.

In [ ]:
gender_p_value = ...
gender_p_value

In [ ]:
grader.check("q3_5")

You should find that the p-value is above the standard cutoff of 0.05 for statistical significance. So in this case, we fail to reject the null. You'll use this information when creating your reboot - you'll assign an equal number of lines to the male main characters and the female main characters.

**Question 3.6.** Conceptually, how would you expect the statistics in `gender_stats` to change if `line_count` were a much larger value and `observed_male_prop` were the same? What effect would that have on the result of the hypothesis test?

From the options below, save the number of your choice in the variable`gender_stats_change`.

1. The values in `gender_stats` would be **less spread out**. We'd be **less** likely to reject the null hypothesis if `observed_male_prop` remained the same.
2. The values in `gender_stats` would be **less spread out**. We'd be **more** likely to reject the null hypothesis if `observed_male_prop` remained the same.
3. The values in `gender_stats` would be **about the same**. We'd be **equally** likely to reject the null hypothesis if `observed_male_prop` remained the same.
4. The values in `gender_stats` would be **more spread out**. We'd be **less** likely to reject the null hypothesis if `observed_male_prop` remained the same.
5. The values in `gender_stats` would be **more spread out**. We'd be **more** likely to reject the null hypothesis if `observed_male_prop` remained the same.

In [ ]:
gender_stats_change = ...
gender_stats_change

In [ ]:
grader.check("q3_6")

<a id='section4'></a>
## Section 4: The Emotional One 😁😢😨

<small>(<a href="#outline">return to the outline</a>)</small>

Now it's time to investigate the character dynamics of the show! In order to do this, we'll look at the `emotions` dataset. In data science, we often refer to emotion as "sentiment." We only have sentiment data for certain lines from the first four seasons of the show, so we'll base our investigation on those lines only. Since we also have knowledge of probability, we'll utilize that as well.

Let's take a look at the data in `emotions`.

In [ ]:
emotions

The `emotions` DataFrame includes the emotion for lines spoken by minor characters like `'Air Hostess'`, but we want to focus on the emotions of the six main characters, whose names are included in the `main` DataFrame defined below.

In [ ]:
main = bpd.DataFrame().assign(speaker =  ['Monica Geller', 'Ross Geller', 'Rachel Green', 'Chandler Bing', 'Phoebe Buffay', 'Joey Tribbiani'])
main

**Question 4.1.** Create a DataFrame called `main_emotions` by merging the `main` and `emotions` DataFrame so that the resulting DataFrame contains the same columns as `emotions`, but only has rows corresponding to lines spoken by one of the six main characters. The rows and columns of `main_emotions` can be in any order.

In [ ]:
main_emotions = ...
main_emotions

In [ ]:
grader.check("q4_1")

We'll use `main_emotions` instead of `emotions` for the rest of this section. Let's look into how sentiment varies across the six main characters by investigating some probabilities.

**Question 4.2.** If we randomly select a line spoken `'Phoebe Buffay'` from `main_emotions`, what is the probability that the line is characterized as `'Joyful'`? Assign your answer to the variable `p_joyful_given_phoebe`.

In [ ]:
p_joyful_given_phoebe = ...
p_joyful_given_phoebe

In [ ]:
grader.check("q4_2")

**Question 4.3.** It's hard to know from this probability alone whether Phoebe is a particularly joyful character compared to other characters. Let's instead answer a related question: if we randomly select a `'Joyful'` line from `main_emotions`, what is the probability that it was spoken by `'Phoebe Buffay'`? Assign your answer to the variable `p_phoebe_given_joyful`.

In [ ]:
p_phoebe_given_joyful = ...
p_phoebe_given_joyful

In [ ]:
grader.check("q4_3")

Notice that in both of the previous questions, you calculated a conditional probability. First you calculated the probability of an emotion given a speaker, then you calculated the probability of a speaker given an emotion. Next, we want to do these same kinds of calculations for _all_ of the main characters and _all_ of the emotions. Let's generalize the code for these calculations so that we can more easily compute conditional probabilities with other conditions.

**Question 4.4.** ⭐⭐ Your job is to implement the function `conditional_probability`. It has two arguments, `find` and `given`, both of which are lists. Let's walk through how this works using an example. Suppose we want to compute the probability that a `'Joyful'` line is spoken by `'Phoebe Buffay'` as we did in the previous question.

- `find` is a list of two elements:
    - The first element in `find` is the column in `main_emotions` that contains the event that we are trying to find the probability of. In our example, this is `'speaker'`.
    - The second element in `find` is the value in the aforementioned column that we're trying to find. In our example, this is `'Phoebe Buffay'`.
- `given` is a list of two elements:
    - The first element in `given` is the column in `main_emotions` that contains the event that is known to be true. In our example, this is `'emotion'`.
    - The second element in `given` is the value in the aforementioned column. In our example, this is `'Joyful'`.

Putting this all together, this means that `conditional_probability(['speaker', 'Phoebe Buffay'], ['emotion', 'Joyful'])` should evaluate to your answer from the previous part (but the `conditional_probability` function should work for any example, not just this one).



In [ ]:
def conditional_probability(find, given):
    '''Returns the conditional probability of an event given a known condition.'''
    ...

# This should evalaute to your answer to Question 4.3. Feel free to try calculating other conditional probabilities.
conditional_probability(['speaker', 'Phoebe Buffay'], ['emotion', 'Joyful'])

In [ ]:
grader.check("q4_4")

**Question 4.5.** Now, use the `conditional_probability` function you just wrote to find some probabilities:
- `p_mad_given_ross`: The probability that a line spoken by `'Ross Geller'` is `'Mad'`.
- `p_monica_given_neutral`: The probability that a `'Neutral'` line is said by `'Monica Geller'`.

In [ ]:
p_mad_given_ross = ...
p_monica_given_neutral = ...

print('P(Mad given Ross) = ' + str(p_mad_given_ross))
print('P(Neutral given Monica) = ' + str(p_monica_given_neutral))

In [ ]:
grader.check("q4_5")

**Question 4.6.** ⭐⭐ Now you want to focus on a particular emotion and order the characters from most likely to say a line of that emotion to least likely. Create a function called `rank_speakers_by_emotion`that takes as input a string `emotion` representing one of the seven emotions, and a boolean `draw_plot`.

If `draw_plot` is `True`, the function should plot a horizontal bar chart where the bar lengths represent the probability that a line of the given emotion is spoken by each of the the six main characters. The total length of the six bars should sum to 1 and the bars should be ordered from longest to shortest. Give your plot an appropriate title.

When `draw_plot` is `False`, the function should not draw a plot.

Regardless of the value of `draw_plot`, the function should return an array of the six main characters, ranked in descending order of their probabilities of saying a line of the given emotion.

For example, `rank_speakers_by_emotion('Sad', True)` should behave as follows.
<img src="images/sad_line.jpg" width = 700>

***Hint:*** You can solve this problem in at least two different ways. One way involves calling your `conditional_probability` function repeatedly with different inputs. Another way doesn't use the `conditional_probability` function at all, but instead uses `groupby`. You can use either approach, or try both for extra practice.

In [ ]:
def rank_speakers_by_emotion(emotion, draw_plot):
    '''Returns an ordered array of main characters, ranked by their conditional probabilities of
    speaking a line of the given emotion. Draws a bar graph when draw_plot is True. '''
    ...

# Try to replicate the graph shown in the example.
rank_speakers_by_emotion('Sad', True)

In [ ]:
grader.check("q4_6")

Now, let's try using our function on all seven emotions.

In [ ]:
seven_emotions = main_emotions.get('emotion').unique()
for emotion in seven_emotions:
    print('Characters in descending order of '+emotion+': \n', rank_speakers_by_emotion(emotion, True), '\n')

Now, you want to create a visualization that will represent the distribution of different emotions for each character. You plan to give this visualization to the writers you will hire for the reboot, as they will need to understand the emotional range of each character to write a reboot that is true to the spirit of the original *Friends* sitcom.

**Question 4.7.** Your task is to create a function `emotions_by_speaker` that will take in a `speaker` and calculate the probability of that character saying a line of each of the seven emotions. The function should return a DataFrame, indexed by `'emotion'`, with one column `'probability'`, containing values that sum to 1. The rows should appear **alphabetically** by `'emotion'`

For example, `emotions_by_speaker('Chandler Bing')` should return this DataFrame:

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>probability</th>
    </tr>
    <tr>
      <th>emotion</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Joyful</th>
      <td>0.220890</td>
    </tr>
    <tr>
      <th>Mad</th>
      <td>0.124429</td>
    </tr>
    <tr>
      <th>Neutral</th>
      <td>0.306507</td>
    </tr>
    <tr>
      <th>Peaceful</th>
      <td>0.096461</td>
    </tr>
    <tr>
      <th>Powerful</th>
      <td>0.080479</td>
    </tr>
    <tr>
      <th>Sad</th>
      <td>0.044521</td>
    </tr>
    <tr>
      <th>Scared</th>
      <td>0.126712</td>
    </tr>
  </tbody>
</table>

In [ ]:
def emotions_by_speaker(speaker):
    '''Returns an DataFrame of emotions and their associated probabilities for the given speaker.'''
    ...

# An example call to your function. Feel free to change this and try out other speakers.
emotions_by_speaker('Chandler Bing')

In [ ]:
grader.check("q4_7")

**Question 4.8.** Now, we'll use your function to create a visualization of speaker emotions, which should show the similarities and differences between the types of emotions each character uses when they speak. We've completed the visualization code for you, you just need to fill in a few missing lines as indicated by the comments. You do not need to know how the code for this visualization works!

In [ ]:
# Set characters to an array of the names of the six main characters.
characters = ...

dfs = {}
for speaker in characters:
    # Call your emotions_by_speaker function on each speaker, and save the result as dfs[speaker].
    dfs[speaker] = ...

def first_name(character):
    # Return the first name of the input character's name, converted to lowercase.
    # For example, RacheL Green -> rachel.
    ...

fig, axes = plt.subplots(2, 3, figsize=(15, 10))
axes = axes.flatten()

max_height = max_width = 0
for character in characters:
    img = plt.imread(f'images/face_pics/{first_name(character)}.png')
    h, w, _ = img.shape
    max_height = max(max_height, h)
    max_width = max(max_width, w)

for ax, character in zip(axes, characters):
    df = dfs[character]
    pie, _, _ = ax.pie(df.get('probability'), labels=df.index, startangle=90, autopct='%1.1f%%', pctdistance=0.85, normalize=False)
    ax.set_title(character)

    img = plt.imread(f'images/face_pics/{first_name(character)}.png')

    h, w, _ = img.shape
    radius = min(h, w) // 2
    y, x = np.ogrid[:h, :w]
    mask = ((y - h // 2) ** 2 + (x - w // 2) ** 2) > radius ** 2
    img[mask] = [1, 1, 1, 0]

    zoom_factor = max_height / h if h > w else max_width / w
    imagebox = OffsetImage(img, zoom=zoom_factor * 0.43)
    ab = AnnotationBbox(imagebox, (0.5, 0.5), frameon=False, pad=0, xycoords='axes fraction', boxcoords='axes fraction')
    ax.add_artist(ab)

plt.tight_layout()
plt.show()

In [ ]:
grader.check("q4_8")

This visualization will serve as a great template for the writers of the new reboot!

<a id='section5'></a>
## Section 5: The One with the Spoilers 🙊

<small>(<a href="#outline">return to the outline</a>)</small>

In this section, we will use permutation testing to compare different groups of episodes from *Friends* in terms of their viewership and ratings. This might help us make some decisions about what kinds of episodes our reboot should include to boost viewership and ratings!

We'll start by looking at how earlier seasons of *Friends* (seasons 1-5) compared to later seasons (seasons 6-10) in terms of viewership. Check out the information you'll be working with in the `episodes` DataFrame:

In [ ]:
episodes

You'll notice that we have data about viewership and rating for 236 episodes (10 seasons). Let's start by labeling the episodes in `episodes` to make it easier to answer our question:
> *Which half of the show has more viewers?*

**Question 5.1.** Create a DataFrame called `halves` which has the same data as `episodes` and an additional column called `'which_half'`. In this column, values should be either `'first half'` (for seasons 1-5) or `'second half'` (for seasons 6-10).

In [ ]:
halves = ...
halves

In [ ]:
grader.check("q5_1")

**Question 5.2.** Before we start permutation testing, let's visualize the difference in viewership between the first half and second half of *Friends* episodes. Create an overlaid density histogram that compares the distribution of `'us_views_millions'` for first half episodes with that of second half episodes. Add a title, legend, and appropriate axis labels to your plot.

Then, assign the difference in **average** viewership (in millions) between the two groups to the variable <code>view_difference</code>. Here, `view_difference` should represent how much **greater** the first half viewership is than the second half, on average.

***Hint:***  Refer to the smoking and birth weight example in [Lecture 22](https://dsc10.com/resources/lectures/lec22/lec22.html#Visualizing-the-distribution-of-each-group) for help creating the overlaid histogram! Think about what bin size is best suited to show the differences between these two distributions. You might want to tweak the bins until the differences become clearer.

<!-- BEGIN QUESTION -->

<!--
BEGIN QUESTION
name: q5_2
points: 1
manual: True
-->

In [ ]:
view_difference = ...
view_difference

In [ ]:
grader.check("q5_2")

<!-- END QUESTION -->



You should see that viewership was higher in the first half of *Friends* than in the second. Is this difference significant or was it just by chance? Let's perform a permutation test to find out.

**Question 5.3.** ⭐⭐ Let's clean up our DataFrame to focus only on the columns `'us_views_millions'` and `'which_half'`. To start, create a DataFrame <code>'views_by_group'</code> that only includes these two columns from `halves`.

We would like to know whether the first half of episodes had significantly **more** viewership than the second half on average. Think about what your null and alternative hypotheses would be to answer this question.

In the previous question, you computed the *observed statistic*, the average number of views by which the first half exceeds the second half. Perform a permutation test, simulating 1000 differences and storing them in the array <code>simulated_differences</code>.

Once you're done, generate a density histogram of the simulated differences, which also shows your observed statistic for comparison. See [Lecture 19](https://dsc10.com/resources/lectures/lec19/lec19.html#Step-3-%E2%80%93-Visualize-the-resulting-distribution) for an example of the type of plot you'll need to create.

In [ ]:
views_by_group = ...
simulated_differences = ...

In [ ]:
grader.check("q5_3")

**Question 5.4.** Calculate a p-value for this permutation test using the observed statistic and the simulated differences, and assign the result to `p_halves`. Then, interpret these results at the 0.05 significance level by assigning <code>True</code> to the variable <code>significant_difference</code> if you believe we can **reject** the null hypothesis and <code>False</code> otherwise.

In [ ]:
p_halves = ...
p_halves

In [ ]:
grader.check("q5_4_a")

In [ ]:
significant_difference = ...
significant_difference

In [ ]:
grader.check("q5_4_b")

We've answered our first question with permutation testing, which helped us figure out whether there was a significant difference in viewership between two groups of episodes. We can use this insight when determining how many episodes our reboot should have. We'll explore the length of our reboot more in Section 7 of this project.

Now, let's ask a different question that can be solved by the same tools:
> *Do people prefer episodes where at least two of the main characters are in a romantic relationship over episodes where there aren't any couples?*

First, we need to label which episodes have an ongoing romantic relationship.

<b><font color='red'>⚠ Spoiler alert! ⚠</font></b>

In the show, there are two couples that each date for a period of time.
 - Ross and Rachel are in a relationship from Season 2, Episode 14 to Season 3, Episode 15, inclusive.
 - Monica and Chandler are in a relationship from Season 4, Episode 24 to Season 10, Episode 18, inclusive, which is through the end of the show.

**Question 5.5.** ⭐⭐ Create a new DataFrame called `relationship` that has the same information as `episodes` and an additional column called `'couple'` that contains `True` if the episode falls in one of the intervals mentioned above, and `False` otherwise.

***Hints:***
- To solve this problem with `apply`, you need to get around the limitation that you can only `apply` a function of one parameter. Consider writing a function that takes strings of the form `'x,y'` where `x` represents the season number and `y` represents the episode number.
- It is also possible to solve this problem without using `apply`. For a challenge, try both ways and see which way is simpler!

In [ ]:
relationship = ...
relationship

In [ ]:
grader.check("q5_5")

**Question 5.6.**  Now, let's state the hypotheses for our permutation test to answer the question:
> *Do episodes where characters **are** in a relationship have higher ratings than those where characters **aren't** in a relationship?*

Consider the following choices for statements of the null and alternative hypotheses. For simplicity, let us call the relationship group "R" (for "relationship") and the non-relationship group "S" (for "single").

1. There is no difference in average episode rating between groups R and S.</li>
1. There is a difference in average episode rating between groups R and S.</li>
1. Group R has a higher rating than Group S, on average.</li>
1. Group S has a higher rating than Group R, on average.</li>

Set `null_choice` to a number 1 through 4, corresponding to your choice for the null hypothesis. Set `alternative_choice` to a number 1 through 4, corresponding to your choice for the alternative hypothesis.

In [ ]:
null_choice = ...
alternative_choice = ...

In [ ]:
grader.check("q5_6")

**Question 5.7.** Using the <code>relationship</code> DataFrame, compute the *observed statistic* representing the difference in mean rating between groups of episodes that have relationships (R) versus those that don't (S), in the order of R minus S. Store your result in the variable <code>couple_observed</code>.

In [ ]:
couple_observed = ...
couple_observed

In [ ]:
grader.check("q5_7")

**Question 5.8.** Now it's time to run the permutation test. Similarly to how you did in Question 5.3, simulate 1000 differences of group means, storing them in the array `couple_stats`. Then, plot your simulated differences in rating and show how they compare to your observed statistic. Make sure you calculate your simulated differences by subtracting in the same order as you did when calculating your observed statistic.

In [ ]:
couple_stats = ...

In [ ]:
grader.check("q5_8")

**Question 5.9.** Calculate your p-value using the observed statistic and simulated differences, storing the result in `p_couple`. Assign <code>True</code> to the variable <code>couple_claim</code> if you believe we can **reject** the null hypothesis at the 0.05 significance level and <code>False</code> otherwise.

In [ ]:
p_couple = ...
p_couple

In [ ]:
grader.check("q5_9_a")

In [ ]:
couple_claim = ...
couple_claim

In [ ]:
grader.check("q5_9_b")

Although we can't determine from our permutation test whether characters being in a relationship *causes* better ratings, we can detect an association. You'll use this information when planning the storyline of your reboot to determine whether to include a romance angle! 💘

<a id='section6'></a>
## Section 6: The One with the Generated Episode Titles 🖨️

<small>(<a href="#outline">return to the outline</a>)</small>

In this section, we'll create a tool that randomly generates episode titles for our reboot, based on the words used in the titles of *Friends* episodes. To start, let's see how *Friends* episode titles are formatted. The cell below saves all the *Friends* episode titles in an array called `titles` and prints a random sample of 10 titles.

In [ ]:
titles = np.array(episodes.get('title'))
for i in np.arange(10):
    print(np.random.choice(titles, replace=False))

As you can see, most episode titles describe a major plot point of that episode and start with a phrase such as `'The One Where'` or `'The One with'`. In other words, the title of an episode is how you might describe the episode to someone else.

When generating episode titles for our reboot, we want to use many of the same words that appear in the titles of *Friends* episodes, so that our reboot titles stay true to the spirit of the original show and follow a similar structure. The simplest way to do this is to randomly select words, one at a time and independently, from the set of words included `titles`.

The variable `every_word` defined below is an array that contains all of the words in all of the *Friends* episode titles.

In [ ]:
every_word = np.array((' ').join(titles).split(' '))
every_word

Some words appear in `every_word` much more frequently than others. We'd want these more common words, like `'The'` to appear more frequently in our generated titles, since they appear more frequently in actual *Friends* titles. Let's generate our titles randomly, one word at a time, so that each word is independently selected with a probability proportional to how often the word appears in a *Friends* title.

**Question 6.1.**  Complete the implementation of the function `find_proportion` which takes as input a word, and returns the proportion of words in `every_word` that are equal to the input word. This should be case-sensitive!

Then calculate the proportion of words that are equal to `'Rachel'` and `'One'`, and save your results in `rachel_prop` and `one_prop`, respectively.

In [ ]:
def find_proportion(word):
    '''Returns the proportion of words equal to the input word.'''
    ...

rachel_prop = ...
one_prop = ...
print('Rachel:', rachel_prop )
print('One:', one_prop)

In [ ]:
grader.check("q6_1")

We can interpret each word's proportion of usages as a probability of that word being selected when we randomly generate a reboot title. For example, if we generate a title by independently selecting three words, we can calculate the probability of generating the title `'Rachel One Rachel'` as follows, using the [multiplication rule for independent events](https://dsc10.com/resources/lectures/lec11/lec11_11am.pdf):

$$
P(\text{Rachel One Rachel}) = P(\text{Rachel}) * P(\text{One}) * P(\text{Rachel})
$$

In [ ]:
prob_rachel_one_rachel = rachel_prop * one_prop * rachel_prop
prob_rachel_one_rachel

That's about a 1 in 50,000 chance that we'll generate the title `'Rachel One Rachel'`, which is so small primarily because `'Rachel'` is a rare word. The title `'One One One'` has a much larger chance of being generated, about 1 in 200, since `'One'` is a common word.

In [ ]:
prob_one_one_one = one_prop ** 3
prob_one_one_one

While the strategy we've outlined gives a way of randomly generating reboot titles so that more common words are more likely to appear in the reboot title, it is unsatisfactory in many ways. Primarily, with this strategy, we can get titles like `'Rachel One Rachel'` and `'One One One'`, which not only don't make sense, but don't follow the structure of *Friends* episode titles that started with phrases like `'The One Where'`.

Instead of considering the probabilities of individual words, let's consider the probability of **pairs** of consecutive words. For example, the pair of words `'The One'` is quite common and the pair of words `'The Last'` is quite rare. This says that when we generate a reboot title randomly one word at a time, if we start with the word `'The'`, we should be more likely to follow it with `'One'` than `'Last'`.

We'll work on calculating the probability distribution of words that immediately follow a given word in *Friends* episode titles. As before, this is case-sensitive. For example, if the word is `'The'`, we'd want to look for want words that follow `'The'` only, not words that follow `'the'`.

**Question 6.2.** ⭐⭐ Complete the implementation of the function `next_word` which takes as input a single episode `title` and a `word`, both strings. The function should return an array containing every word in `title` that appears immediately after `word`, or an empty array if there are no such words. This could happen either because the `word` is the last word in `title`, or because the `word` is not a part of `title` at all. The output array can contain duplicates if the same word follows the input `word` multiple times.

Example behavior is shown below.

```py
>>> next_word('The One Where the Monkey Gets Away', 'Monkey')
array(['Gets'])

>>> next_word('The One with the Chicken Pox', 'Pox')
array([])

>>> next_word('This is the final project of the course', 'the')
array(['final', 'course']

>>> next_word('This is the final project of the course and there is also the final exam', 'the')
array(['final', 'course', 'final']


```


In [ ]:
def next_word(title, word):
    '''Given a title and an input word, return an array of all
    words in the title that immediately follow the input word.'''
    ...

# An example call to your function. Feel free to change this and try out other inputs.
next_word('The One Where the Monkey Gets Away', 'Monkey')

In [ ]:
grader.check("q6_2")

**Question 6.3.** Now that we know how to find the subsequent words in a single title, let's extend what we've done to find the subsequent words across all titles in `titles`. Complete the function `words_following`, which takes as input a single `word` and returns an array of all words that immediately follow `word` in a *Friends* episode title. To do this, you should call your `next_word` function once on each title.

As before, the output array can contain duplicates if the same word follows `word` multiple times, and it can be empty if no word follows `word`.

***Hint:*** You can use `np.append` to append a whole array of values onto an existing array, in the same way that you would append a single value onto an existing array.

<!--
BEGIN QUESTION
name: q6_3
points: 1
-->

In [ ]:
def words_following(word):
    '''Given an input word, return an array of all words that
    immediately follow the input word in episode titles in titles.'''
    ...

# An example call to your function. Feel free to change this and try out other input words.
words_following('One')

In [ ]:
grader.check("q6_3")

Now that we know how to find all the words that follow a given word, we can create a probability distribution, where the probability of each subsequent word is just the proportion of times that word follows the given word.

**Question 6.4.** ⭐⭐ Complete the `find_prob_distribution` function which takes as input a single `word` and returns a DataFrame indexed by subsequent words that can immediately follow `word`, and with one column called `'prob'` that contains the proportion of times the subsequent word follows `word`. Sort the rows in descending order of `'prob'` so that the most likely words are at the top of the DataFrame.

For example, `find_prob_distribution('The')` should look like this:

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>prob</th>
    </tr>
    <tr>
      <th>word</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>One</th>
      <td>0.987288</td>
    </tr>
    <tr>
      <th>Last</th>
      <td>0.008475</td>
    </tr>
    <tr>
      <th>Pilot</th>
      <td>0.004237</td>
    </tr>
  </tbody>
</table>

<!--
BEGIN QUESTION
name: q6_4
points: 2
-->

In [ ]:
def find_prob_distribution(word):
    '''Returns a DataFrame containing the probability distribution
    of words that can follow the given input word.'''
    ...

# An example call to your function. Feel free to change this to try out other input words.
find_prob_distribution('The')

In [ ]:
grader.check("q6_4")

Now, we can use these probability distributions to generate titles randomly. We'll start each title with the word `'The'`, since all *Friends* episodes start with this word. Then, the next word will be `'One'` with probability 0.987288, or `'Last'` with probability 0.008475, or `'Pilot'` with probability 0.004237. Almost certainly, the next word will be `'One'`, and then we'll figure out the next word according to the probabilities in the DataFrame below.

In [ ]:
find_prob_distribution('One')

Perhaps by random chance we'll end up with the next word being `'Proposes'` (though it's quite unlikely). To find the words that could follow `'Proposes'` along with their associated probabilities, we'll use this DataFrame:

In [ ]:
find_prob_distribution('Proposes')

Since this DataFrame has no rows, it means `'Proposes'` never appears in a *Friends* episode title with a word after it. That is, it only appears as the last word in an episode title. This means no word should come after it, so we're done generating our title, which in this example was `'The One Proposes'`.

When we generate titles based on pairs of words in this way, the probability of generating `'The One Proposes'` can be computed as:

$$
P(\text{The One Proposes}) = P(\text{The}) * P(\text{One}|\text{The}) * P(\text{Proposes}|\text{One})
$$

Since all episode titles start with `'The'`, $P(\text{The}) = 1$. The other probabilities are conditional probabilities based on the previous word. For example, $P(\text{One}|\text{The})$, which we read as "the probability of `'One'` given `'The'`", represents the probability of seeing the word `'One'` immediately after the word `'The'`. According to our `find_prob_distribution` function, this probability is quite high at 0.987288.

In [ ]:
find_prob_distribution('The') # This says P(One|The) = 0.987288.

Similarly, we can find $P(\text{Proposes}|\text{One})$ is very small, only 0.004274.

In [ ]:
find_prob_distribution('One') # This says P(Proposes|One) = 0.004274

Therefore, the probability of generating `'The One Proposes'` is:

$$
\begin{align}
P(\text{The One Proposes}) &= P(\text{The}) * P(\text{One}|\text{The}) * P(\text{Proposes}|\text{One}) \\
&= 1 * 0.987288 * 0.004274 \\
&\approx 0.004219
\end{align}
$$

This is about 1 in 2000.

**Question 6.5.** Now that we know how to generate titles based on pairs of words, let's write a function that takes no inputs, and returns a randomly generated title by starting with the word `'The'`, then adding a new word based on the conditional probabilities of words that can follow `'The'`. We'll continue adding new words after each previous word, until we encounter a word that never had any words after it in an episode title (like `'Proposes'` in our example above). At that point, we'll stop generating new words.

Since we don't know how many words we'll be adding in advance, it's hard to implement this with a `for`-loop, since we don't know how many iterations we'll need. Instead, we'll use what's known as a `while`-loop, which runs continuously, with as many iterations as needed, until some stopping condition is met. In our case, we'll keep adding new words as long as the DataFrame returned by `find_prob_dist` has at least one row, since that means there are more words we can add to our title. Our stopping condition, therefore, is when `find_prob_dist` produces a DataFrame with no rows.

We've implemented the `while`-loop for you. Your task is to fill in the missing lines so that `generate_title` returns a randomly generated episode title. Some comments are included to help you.

<!--
BEGIN QUESTION
name: q6_5
points: 1
-->

In [ ]:
def generate_title():

    generated_title = 'The'
    prob_dist_df = find_prob_distribution('The')

    # Keep generating new words for our title so long as there are
    # words that can follow the most recent word we've added.
    while (prob_dist_df.shape[0] >= 1):

        # Set new_word to a random selection of the words represented in prob_dist_df,
        # chosen according to their probabilities in prob_dist_df.
        new_word = ...

        # Add onto your generated title in the variable generated_title.
        # Make sure to include spaces between words.
        generated_title = ...

        # Update prob_dist_df so that it contains the probability
        # distribution of the word you just added to your title.
        prob_dist_df = ...

    return generated_title

generate_title()

In [ ]:
grader.check("q6_5")

**Question 6.6.** Suppose you generate one title using the `generate_title()` function above. What is the probability that the title you generate is `'The One in Vegas'`? Set your answer to `vegas_prob`.

***Note:*** This is a probability question that you should answer based on the conditional probability of seeing one word given the previous. You should not do any simulation for this question.

In [ ]:
vegas_prob = ...
vegas_prob

In [ ]:
grader.check("q6_6")

**Question 6.7.** You want to know how frequently your `generate_title()` function generates a title that actually is a real title for a *Friends* episode. Based on a simulation with 1,000 trials, estimate the probability of a generated title being a real *Friends* title, and store the result in `prob_real`.

In [ ]:
prob_real = ...
prob_real

In [ ]:
grader.check("q6_7")

You should find that there's a fairly high probability of generating a title that already exists. This happens because there aren't really that many words in the set of *Friends* episode titles. For example, if we randomly generate the word `'East'`, it's going to have to be followed up with `'German Laundry Detergent'` because the word `'East'` only appears once, which is in the episode title `'The One with the East German Laundry Detergent'`. With a relatively small body of text in *Friends* episode titles, there are many words that can only be followed by one thing.  For simplicity, we'll ignore the fact that we are generating many existing episode titles, though we could significantly improve the situation by basing our titles not just on past *Friends* episode titles, but on a larger body of text, such as the full scripts of all *Friends* episodes. This improvement would come at the cost of much more computation!

Now, let's put all our hard work to use and generate episode titles for the first season of the reboot. Most seasons of *Friends* had 24 episodes, so you'll make sure your reboot starts out with a season of 24 episodes. Run the cell below to generate the titles for the first season of the reboot!

In [ ]:
for i in np.arange(1, 25):
    print('Reboot Episode', i, ':', generate_title() )

In this section, you learned how to use pairs of words to generate text. More generally, sequences of $n$ words (called $n$-grams) can be used to generate text in a similar way. For example, if $n=3$, we would base each next word on the conditional probability of that word appearing after the previous two words. To start, we'd consider what words follow the phrase `'The One'` and how frequently. Perhaps we'd randomly choose the next word to be `'Where'`. We'd choose our next word based on words that follow `'One Where'`, and so on.

$n$-grams are used to capture the patterns and sequences of words in a language, which can be useful for a variety of natural language processing tasks, such as language modeling, text generation, machine translation, and more.  While this is a simplified example, modern language models like GPT (Generative Pre-trained Transformer) use more advanced neural network architectures and techniques to capture long-range dependencies between words and generate more coherent and contextual text.

<a id='section7'></a>
## Section 7: The Last One 🔚🥳

<small>(<a href="#outline">return to the outline</a>)</small>

In this section, you'll determine how many episodes your reboot should have to be successful. The first question is how should you measure success, in terms of views or in terms of ratings? You suspect that if an episode is good, it’s only natural that it will get more views *and* higher ratings, right? So there should be a positive correlation between viewership and ratings. You decide to explore the connection between these two variables using regression.

Let's start by visualizing the data with a scatter plot to see if linear regression would make sense for this dataset.

In [ ]:
episodes.plot(kind='scatter', x='imdb_rating', y='us_views_millions');

Based on the scatter plot, it seems like there is a slight positive linear association, and so linear regression would be an appropriate tool. Let's continue!

**Question 7.1.** The first step will be to convert our data into standard units. Complete the function `standard_units` which takes in an array or Series of numerical data and returns an array with the values in standard units. Then use your function to standardize the `'imdb_rating'` and `'us_views_millions'` columns from `episodes`. Store the standardized arrays in the variables `rating_standard` and `views_standard`.

In [ ]:
def standard_units(sequence):
    '''Returns the input sequence as an array in standard units.'''
    # Convert the input to an array, if it is not already.
    sequence = np.array(sequence)
    ...

rating_standard = ...
views_standard = ...

In [ ]:
grader.check("q7_1")

**Question 7.2.** Now that we know how to convert variables to standard units, we can calculate the correlation. Complete the function `correlation`, which should take in:
1. `df`, a DataFrame,
2. `independent`, the column label of the independent ($x$) variable, as a string, and
3. `dependent`, the column label of the dependent ($y$) variable, as a string.

The function should output the correlation between the two variables.

Then, use your function to compute the correlation between `'imdb_rating'` and `'us_views_millions'` and store your result in the variable `corr`. Check that the number you get looks reasonable based on the scatter plot above.

In [ ]:
def correlation(df, independent, dependent):
    '''Returns the correlation between the independent and dependent variables in the given DataFrame.'''
    ...

corr = ...
corr

In [ ]:
grader.check("q7_2")

**Question 7.3.** Now construct two functions, `reg_slope` and `reg_intercept`, which each take in the same three inputs as `correlation`. `reg_slope` should return the slope of the regression line and `reg_intercept` should return the intercept of the regression line, in original units.

Use your function to store the slope and intercept of the regression line for  `'imdb_rating'` and `'us_views_millions'` in the variables `slope` and `intercept`.

In [ ]:
def reg_slope(df, independent, dependent):
    '''Returns the slope of the regression line in original units.'''
    ...

def reg_intercept(df, independent, dependent):
    '''Returns the intercept of the regression line in original units.'''
    ...

slope = ...
intercept = ...
slope, intercept

In [ ]:
grader.check("q7_3")

**Question 7.4.** Create a function called `predict` that takes in the same three inputs as `correlation`. `predict` should return an array of predicted values of the dependent variable calculated from the regression line.

Use your function to create an array of the predicted number of views, in millions, for each episode in the `episodes` DataFrame, based on the episode's rating. Save your answer as `predicted_views`.

In [ ]:
def predict(df, independent, dependent):
    '''Returns an array of predicted values of the dependent variable calculated from the regression line.'''
    ...

predicted_views = ...
predicted_views

In [ ]:
grader.check("q7_4")

**Question 7.5.** Use the strategy for overlaying scatter plots described in [this tutorial](https://www.statology.org/pandas-scatter-plot-multiple-columns/) to create an overlaid scatter plot with:
- a blue dot 🔵 for each episode showing the ratings on the $x$-axis and the views on the $y$-axis (as in the scatter plot at the beginning of this section), and
- a red dot 🔴 for each episode showing the ratings on the $x$-axis and the **predicted** views on the $y$-axis.

The red dots should form a straight line - that's the regression line!

***Note:*** This is the first time you've been asked to make an overlaid scatter plot, so you'll need to learn something new to answer this question. Read the linked tutorial carefully and try to mimic their example; you can learn how to do a lot of things this way!

<!-- BEGIN QUESTION -->

<!--
BEGIN QUESTION
name: q7_5
points: 1
manual: True
-->

In [ ]:
# Create your overlaid scatter plot here.
...

<!-- END QUESTION -->



**Question 7.6.** Use the equation of the regression line to answer the following questions. Check that your answers are reasonable using the scatter plot above.

1. What is the predicted number of views, in millions, for an episode with an IMDb rating of 8.6? Save your answer as `question_1`.
2. An episode is predicted to have 30 million US views. What is its IMDb rating? Save your answer as `question_2`.
3. Suppose an episode in season one has an IMDb rating of $d$ and an episode in season two has a IMDb rating of $d + 2$. How many millions more views would we predict the episode in season two to have compared to the episode in season one? Save your answer as `question_3`.


In [ ]:
question_1 = ...
question_2 = ...
question_3 = ...
print('Answer to question 1: ', question_1)
print('Answer to question 2: ', question_2)
print('Answer to question 3: ', question_3)

In [ ]:
grader.check("q7_6")

**Question 7.7.** Now that we have general code to calculate the regression line between any pair of variables in any DataFrame, let's generalize our code for the overlaid scatter plot so we can visualize relationships between other pairs of variables.

Complete the function `display_predictions` below. This function should take in the same three inputs as the `correlation` function, create an overlaid scatter plot similar to the one in Question 7.5, and return a string describing the correlation between the variables and the slope and intercept of the regression line.

Then, use your function to create a scatter plot and calculate the regression line that would allow you to predict IMDb rating based on viewership in millions. Store the result of your function call in the variable `rating_prediction`.

In [ ]:
def display_predictions(df, independent, dependent):
    '''Generates an overlaid scatter plot showing the relationship between the independent and dependent variables in df.
    Returns a string describing the correlation and the slope and intercept of the regression line.'''
    # Create your overlaid scatter plot here.
    ...

    # We've provided the code for the return statement.
    return ('The correlation between {0} and {1} is {2}. ' +\
           ' The slope of the regression line is {3}.' + \
           ' The intercept of the regression line is {4}.')\
                .format(independent,
                        dependent,
                        str(round(correlation(df, independent, dependent), 2)),
                        str(round(reg_slope(df, independent, dependent), 2)),
                        str(round(reg_intercept(df, independent, dependent), 2)))

# Your function should produce the same scatter plot as in Question 7.5 on the inputs below.
# Make sure to test it out on other inputs too.
display_predictions(episodes, 'imdb_rating', 'us_views_millions')

In [ ]:
rating_prediction = ...
rating_prediction

In [ ]:
grader.check("q7_7")

**Question 7.8.** There seem to be just a few episodes with 40 million views or more. Since outliers can have a big impact on the regression line, let's see how different our regression line would look if we eliminated these outliers. Again using the `display_predictions` function you just wrote, create a scatter plot and calculate the regression line that would allow you to predict IMDb rating based on viewership, but based only on episodes with less than 40 million views. Store the result of your function call in the variable `no_outliers`.

In [ ]:
no_outliers = ...
no_outliers

In [ ]:
grader.check("q7_8")

Let's use some of the regression tools we've developed in this section to answer an important question about the reboot: how many episodes should it have? You are about to meet with the producers of the reboot to discuss this. To inform your discussion, you want to see how viewership of *Friends* changed as more episodes were made.

**Question 7.9.** Call your `display_predictions` function so that it displays the regression line predicting `'us_views_millions'` based on the number of *Friends* episodes produced.

***Hint:*** The `episodes` DataFrame does not have a column that counts the episode number from 1 to the total number of episodes. You'll need to add one!

<!-- BEGIN QUESTION -->

<!--
BEGIN QUESTION
name: q7_9
points: 1
manual: True
-->

In [ ]:
...

<!-- END QUESTION -->



Since the slope of the regression line is negative, this means that viewership of *Friends* ultimately declined as more episodes were created. But this wasn't always the case. When the show was just starting off, new episodes were getting more views than past episodes. The regression line fit to just the first 30 episodes, for example, would have a strongly positive slope.

You interpret this data as meaning that maybe *Friends* went on for too long. You want your reboot to have few enough episodes that the excitement of the show doesn't wear off.

**Question 7.10.**  Use your `display_predictions` function to display the regression lines predicting `'us_views_millions'` based on the first 20, 40, 60, 80, 100, and 120 episodes. The largest such number of episodes with a positive slope is the number of episodes you think the reboot should have. Save this number of episodes as `reboot_length`. You can type this number in manually after looking at the plots.

In [ ]:
# Plot your regression lines here.
...

reboot_length = ...
reboot_length

In [ ]:
grader.check("q7_10")

Now that you know how many episodes your reboot should have, you've completed your plans for the reboot. Well done and best wishes for a successful show!

## Finish Line 🏁

### **Congratulations! 🎉 You've completed the Final Project!**

**_Citations:_** Did you use any generative artificial intelligence tools to assist you on this assignment? If so, please state, for each tool you used, the name of the tool (ex. ChatGPT) and the problem(s) in this assignment where you used the tool for help.

<hr style="color:Maroon;background-color:Maroon;border:0 none; height: 3px;">

Please cite tools here.

<hr style="color:Maroon;background-color:Maroon;border:0 none; height: 3px;">

To submit your assignment:

1. Select `Kernel -> Restart & Run All` to ensure that you have executed all cells, including the test cells. <p style="color: red"><b>⚠️ Important!</b> We will allot 20 minutes of computer time to run your notebook. If your notebook takes longer than this to run, it may not pass the autograder! Select "Kernel -> Restart and Run All" to time how long your notebook takes. A notebook with correct answers should take less than 10 minutes.</p>

2. Read through the notebook to make sure everything is fine and all tests passed.
3. Run the cell below to run all tests, and make sure that they all pass.
4. Download your notebook using File -> Download as -> Notebook (.ipynb), then upload your notebook to Gradescope. Don't forget to add your partner to your group on Gradescope!
5. Stick around while the Gradescope autograder grades your work. Make sure you see that all tests have passed on Gradescope.
6. Check that you have a confirmation email from Gradescope and save it as proof of your submission.

If running all the tests at once causes a test to fail that didn't fail when you ran the notebook in order, check to see if you changed a variable's value later in your code. Make sure to use new variable names instead of reusing ones that are used in the tests.

Remember, the tests here and on Gradescope just check the format of your answers. We will run correctness tests after the assignment's due date has passed.

In [ ]:
grader.check_all()

<img style="float: left;" src="images/friends_anticipation.png" width="600" height="600">